### ISSUE TITLE
Native `VAR` syntax to create variables inside tests and keywords

### ISSUE THREAD

**Issue by mkorpela** [robotframework/robotframework#3761]
Robot Framework row is parsed by the following rules:
- First column is thought as a keyword definition.
- Unless if first column is a variable then second column is thought as a keyword definition.
- Unless if the second column is also a variable then apply second rule again.
- All the rest columns are paramenters to that keyword and by default they are strings (that can be converted to types).

Those rules result in an odd way of setting a simple value for a local variable:
`${my_variable}=  Set Variable  MY_VALUE`

I propose we introduce a shortcut for this to reduce unnecessary verbosity:
`${my_variable}:=  MY_VALUE`

**Comment by xylix** [comment]
I would also consider fish shell style `Set  my_variable  value` syntax. So just keyword `Set` for setting a variable to a value. Wouldn't require any special syntax.

But this `:=` also seems like a significant upgrade.

**Comment by Snooz82** [comment]
This could also just be an addition to BuiltIn library.

I am totally with you that `Set Variable` is a bit unhandy.
But a Keyword with the name `=` or `:=` in BuiltIn would maybe just do the trick.

```
Test
    ${var}    =    Value
    ${list}    =     one    two    three
    ${dict}    =    key=val    key2=val2
    ${eval}     =    ${{sys.argv[1]}}
```


That would maybe just easier to understand and the rules and an logs could stay the same.
If it is added to BuiltIn it is like a language feature.

**Comment by pekkaklarck** [comment]
I agree creating local variables using a hard coded value is a bit annoying and coming up with something nice than

    ${var} =    Set Variable    value

would be nice.

Is @mkorpela your idea that

    ${var} := value

would be a single token i.e. just one space around `:=`? If yes, that would have a problem that it couldn't be lexed using the current general parsing rules and e.g. LSP needed to handle it separately. It wouldn't be compatible with how execution is logged either so it would require quite a lot of changes. I'd also use just `=` instead of `:=` (with required spaces around) because Python uses the "walrus operator" for assignment expression syntax and it behaves a bit differently.

The idea to use

    Set    var    value

by @xylix is actually already mostly supported by the newish

    Set Local Variable    ${var}    value

but obviously this is a bit more verbose syntax. I still don't think adding yet another keyword would help and using something as generic as `Set` would also likely conflict with many keywords users have in their libraries.

The approach proposed by @Snooz82 to allow using `=` as a keyword name like 

    ${var}    =    value

is actually something I've been thinking about earlier. It would have a great benefit that we just needed to make `=` an alias for `Set Variable` (easy!) and then everything else ought to work just fine. Well, probably in some string representations and in the log file we needed to make sure that the assignment would look like `${var} = value` and not like `${var} = = value` (i.e. the `=` automatically added with assignments shouldn't be added). A related problem would be that you could then in the data have

    ${var} =    =    value

but I don't consider that a big problem. You just don't need to use such stupid syntax. Another problem is that

    ${var} =    xxx
    ${var}    =    xxx

would look pretty similar but in the first case `xxx` is a keyword and in the latter case it's a value to assign directly. I don't think that's a big problem either.

If others agree, I'd be fine going with the above syntax and consider adding it already in RF 4. There would still require some design decisions related to how to handle lists and dicts but if that gets complicated just supporting the scalar assignment would already help a lot.

**Comment by pekkaklarck** [comment]
This won't be in RF 4.0 but let's see could it be added in RF 4.1. Still time to think about the syntax as well.

**Comment by pekkaklarck** [comment]
Too big change for a small release like RF 4.1. Moving to RF 5.0.

**Comment by boakley** [comment]
Would it make sense to make `:=` an alias for `Set Variable`?

```
*** Test cases ***
Example
    ${message}  :=  Hello, world
    ${this}  ${that}  :=  string one  string two
```

**Comment by pekkaklarck** [comment]
No consensus on the syntax to use. Moving to RF 5.1 to give us more time to think about this.

We have also been thinking about inline keyword call syntax (#3187) and it should be taken into account when designing this syntax. These two features are interestingly related. A problem with variable assignment syntax is that in `${x} =    something` syntax that `something` is a keyword and we'd like it to be a literal value. With the inline keyword call syntax the problem is opposite: in `something    value` that `value` is an argument and we wanted it to be a keyword.

**Comment by pekkaklarck** [comment]
Using `=` or `:=` as keywords like

**Comment by pekkaklarck** [comment]
One problem with using `=` or `:=` as a keyword like
```
${var}    =    literal value
```
or
```
${var}    :=    literal value
```
is that it wouldn't work nicely with the new inline IF syntax (#4093). The new syntax supports assignment but this would be rather ugly:
```
${var}    IF    ${expr}    =    true value    ELSE    =    false value
```
A much better syntax would be
```
${var}    =    IF    ${expr}    true value    ELSE    false value
```
but now the whole IF/ELSE is an argument to `=` keyword. I guess we could handle `=` already at parsing time so that this could be handled somehow nicely, but it would nevertheless require some more work.

**Comment by pekkaklarck** [comment]
Another thing to take into account is setting variables in different scopes. As #4177 demonstrates, current `Set Global/Suite/Test/Local Variable` keywords have some quirks that are hard to fix, and having some better syntax for that usage would allow avoiding these keywords altogether.

If we'd come up with some nice syntax for creating local variables, we could then add `GLOBAL` and `SUITE` statements that could be used with it. For example, if we'd use `=` as a keyword, then
```
GLOBAL    ${x}    =    literal value
```
could possibly be used.

If we'd use `Set` keyword that takes name and value as arguments, then the scope could be just an optional argument:
```
Set    $x    value    scope=GLOBAL
```

**Comment by pekkaklarck** [comment]
There have been various proposals so far and I have some new ones in my mind as well. I try to summarize different options here:

1. New keyword to replace `Set Variable` like `=` or `:=`. Would be easy to implement, but [as I wrote earlier](https://github.com/robotframework/robotframework/issues/3761#issuecomment-732962895) it would be a bit confusing that
   ```
   ${var} =    Keyword
   ```
   and
   ```
   ${var}    =    literal value
   ```
   look so similar. Not working nicely with inline IF (see comment a bit above) is a minus as well.

2. Some markers around (or possibly only before) the value to separate it from keywords. Below are some alternatives:
   ```
   ${var} =    /literal value/ 
   ${var} =    "literal value"
   ${var} =    <literal value>
   ${var} =    << literal value
   ```
   This syntax would work nicely with inline IF:
   ```
   ${var} =    IF    ${expr}    <true value>    ELSE    <false value>
   ```
   It wouldn't work that nicely with `@{list}` and `&{dict}` variables but I guess this would be ok:
   ```
   @{var} =    <first    second   third>
   ```
 
3. Some marker as part of the assigned variable to tell the value is to use as a literal string. Alternatives:
   ```
   ${var} <    literal value
   ${var} <=    literal value
   ```
   To make this work with inline IF would require some work.

4. New `Set` keyword that explicitly takes name and value:
   ```
   Set    $var    literal value
   ```
   Should require name to be in `$name` format, not `${name}`, to avoid problems with `Set Global/Suite/Test/Local Variable` (#4177) and, as I mentioned in the above comment, could accept `scope` as an optional argument to make these keywords redundant:
   ```
   Set    $var    literal value    scope=GLOBAL
   ```
  Wouldn't work that well with inline IF
   ```
   IF    ${expr}    Set    $var    true value    ELSE    Set    $var    false value
   ```
   but perhaps the keyword itself could support conditional usage so that
   ```
   Set    $var    IF    ${expr}    true value    ELSE    false value
   ```
   would work.

   A major problem is that it is _very_ likely that people have their own keywords named `Set`. If you'd have such a user keyword, it would override the BuiltIn keyword silently, and if you have a library keyword there would be a warning about conflicting names. `Set Variable` would be a better name in that regard, but we already have that keyword and changing its behavior would be too badly backwards incompatible. I'd prefer a shorter name for a keyword that's used this often in general.

5. New `SET` statement that would work like the `Set` keyword discussed above but would be part of RF syntax, not a keyword. Above examples using `Set` would look like this with `SET`:
   ```
   SET    $var    literal value
   SET    $var    literal value    scope=GLOBAL
   IF    ${expr}    SET    $var    true value    ELSE    SET    $var    false value
   SET    $var    IF    ${expr}    true value    ELSE    false value
   ```
   I like how `SET` stands out better than `Set`. It would even work with the inline IF pretty well, but `SET` could nevertheless get its own conditional assignment support. The biggest win over `Set` would be that `SET` statement couldn't be overridden by any keyword. Naturally someone could have used their `Set` keyword like `SET` and now that wouldn't work anymore, but such usage ought to be rare and easily changed.

**Comment by emanlove** [comment]
Some thoughts from this thread ..

Using single token i.e. just one space around := would seriously complicate the "you need two or more spaces" mantra that is drummed into every new user. Adding a single space as a linguistic syntax to me is a VERY, VERY, BAD idea.

With making `:=` an alias for `Set Variable` where it is "tokenized" or separated out by the two spaces on each side I think still has the drawback of a new syntax that I am not sure we benefit from as greatly as it complicates with an ever expanding language.

I think just `=` as in `${var}    =    value` is less of a stretch in terms of not a new symbol in the syntax but a re- or new application of an existing symbol. I do see the complications both @pekkaklarck points out ("'=' in the data" and "the visual similarities but syntax differences between '${var} =    xxx' versus '${var}    =    xxx'" and "how it doesn't work nicely with the new IF ELSE syntax") and @MoreFamed notes ("simple '=' used for assigning is confusing, since people know equal sign as a comparison operator"). So I see it as an option but not a very strong one.

Using markers (like `${var} =    /literal value/`, `${var} =    "literal value"`, `${var} =    <literal value>`) to me is more syntax with little benefits and larger end-user educational barrier.

As I finish going through the thread I am at a lost as to what the goal is? Is it just to replace typing out `Set Variable` as in `${my_variable}=  Set Variable  MY_VALUE`? The drawbacks of this syntax I see are it is verbose and takes typing. But my response is that Robot Framework is a natural language keyword driven framework so I expect some verbosity. And if `Set Variable` is too much typing can't we just add a keystroke to our editors to have it type it out for us? I, too, see the syntax could be "better" but so far no of these alternatives seem to be worth the effort.

**Comment by pekkaklarck** [comment]
The main problem with current `Set Variable` is verbosity. I also don't like using keywords for common syntax now that we have IF/ELSE, TRY/EXCEPT, CONTINUE, BREAK and RETURN as well. I agree using special characters have a problem of being too terse and hard to understand if you don't know Robot Framework well. I'm starting to like the idea of adding SET statement more and more. It would be pretty explicit without being overly verbose. As a bonus it could easily support also global and suite scopes.

**Comment by MoreFamed** [comment]
Thanks. @pekkaklarck, for the analysis. My opinion:
- 1 is my favourite. :) Some remarks: Not everyone uses that unnecessary "=" in assignments, we write just `${variable}  Keyword`. Replacing _Set Variable_ with `:=` would cause much less confusion than with `=`. 
- 2 and 3 are rather technical notations. RF has (should have) a human-readable language, and those technicalities look ugly.
- 4 and 5 looks good, however, I suppose that variables being arguments of the keyword will need to be in special syntax not to be replaced with their values. This can be very confusing not only for beginners! If you will manage to do it in such a way that writing 
```
SET    ${var}    literal value
``` 
would be possible (it will be in a statement, wouldn't it?), it would seem reasonable to me.

I would not care about inline IF much (in all the cases): when someone uses this all-purpose construction for a specific task of conditional assignment to the only-and-always-the-same variable, some ineffeciency or ugliness is expected. I would prefer _Set Variable If_ for this task.

**Comment by JockeJarre** [comment]
great summary @pekkaklarck !

I wanted to add my 5 cents to the mix.

When comparing 

`
SET    ${var}    literal value
`
VS

`
SET    $var    literal value
`

I understand your point why the second version should be used, but I really see an end-user being confused when to use the brackets and when not. For python statements, it is understandable, but here I personally think it get's to complicated.

Rather I believe    `SET    ${var}    literal value`   should never evaluate the value of the variable to try to find out a name, at least not as default behavior. Rather adding a named variable telling that you WANT it to evaluate to a name could be of interest?

like:
`
SET    ${var}    literal value    scope=GLOBAL   evaluate_name=False
`

**Comment by pekkaklarck** [comment]
I agree that requiring
```
SET    $name    value
```
syntax can be confusing and being able to use
```
SET    ${name}    value
```
would be better. I still believe that being able to give the name of the variable to be set itself as a variable is a useful feature. On the Slack discussion I proposed, similarly as @JockeJarre above, that you needed to use
```
SET    ${name}    value    resolve=True
```
if you want the variable to be resolved. In another discussion there it was proposed to use `${${name}}` syntax with `Set Suite Variable` to avoid problems explained in #4177. I don't think we can change that keyword, but the possibly added new statement could require using
```
SET    ${${name}}    value
```
when you want the name to be a variable. The above would work so that `${name}` would be resolved first and then its value would be used as the base name with `SET`. It would avoid the need for a separate configuration parameter and could perhaps be easier to understand as well.

Both using `resolve=True` and `${${name}}` would have nice benefits:
- Just `${name}` could be used in the common case.
- It would be possible to set name based on variable.
- The behavior would be explicit (i.e. we wouldn't have the same problem as in #4177).

**Comment by xylix** [comment]
I would personally find `resolve=True` simpler to read / less confusing than `${${name}}`.

**Comment by Snooz82** [comment]
I have not seen any proposal that could lead to common agreement. 

And the only issue with `Set Variable` keyword is, that it is a keyword. 
because every user can make their own keyword `=` or `:=` or `set` or `assign` or what so ever.

the only benefit of an Inbuilt language feature would be, that it would be nicer to read in the logs.

and to be honest, I personally doubt very much that it is really worth the effort.

**Comment by SamMaksymyshyn** [comment]
There is an idea of a partial solution to overcome the verbosity of the `set variable` name. 
The `is` synonym could be added for it to the Builtin.
E.g. 
```robot
${actual_total}   is   ${table cells[0].text}
```

under the hood
```python
def is_(value):
    return value
```

**Comment by robinmackaij** [comment]
As just discussed in the Open Space:
I would suggest using `VAR` instead of `SET` since many languages use a `var` directive that would be more recognizable to many users.

**Comment by pekkaklarck** [comment]
`VAR` instead of `SET` is a good idea. We could even consider using `VARIABLE` to make it more explicit and easier to possibly translate in the future, but the longer name may also be distracting. How do others feel about this?

**Comment by pekkaklarck** [comment]
Below is my proposal how this should work. I now use `VAR` as the marker, but it can be also `SET`, `VARIABLE`, or something else.
```
# Basic usage, create a local variable with name `${name}` and literal value `value`.
VAR    ${name}    value

# Same as the above, but the value is got from variable `${value}`.
VAR    ${name}    ${value}

# Create a variable in some different scope.
VAR    ${name}    value    scope=GLOBAL

# Resolve the variable name. Variable `${name}` should contain the name of the variable to create.
VAR    ${name}    value    resolve_name=True
```

If we can agree about the above general syntax, I'm sure we can implement this in RF 7 scope. There are, however, some open design decisions to be covered:

1. Marker name. I start to like `VAR` more and more.
2. Should we support both `$name` and `${name}`? I believe we should at least if we implement #4674.
3. Should we use `resolve_name` or just `resolve` when the name should be resolved? The latter is shorter but the former is more explicit and thus probably better. We already have config options with multiple parts like `on_limit_message` with other constructs.

**Comment by MoreFamed** [comment]
@pekkaklarck 

1. I have no strong preference, but I hope the new style of setting variables will be possible to use in inline IFs (unlike `Set Variable` 🤦‍♂️) But something like `VAR` is better than `SET` because when used without value (do you count on that possibility?), `VAR  ${name}` (just to create the empty variable) makes sense semanticaly, while `SET  ${name}` does not (set `${name}` to what?)

1. I agree with you.

1. I would slightly oppose having the argument at all. The `${${var}}` syntax is used widely in RF, I guess, so it can be used here as well. With `resolve=True`, this will be the only place where `${var}` would not mean the variable *var* or its content, but the variable name of which is stored in *var*.

**Comment by pekkaklarck** [comment]
Yeah, we definitely need to support setting variables conditionally. We can either make sure `VAR` (or `SET`, or whatever) works with inline IF so that you can use

**Comment by emanlove** [comment]
Just my two cents .. (and a bit late). I prefer to leave it as `Set Variable`. For me it is going down that path of too much like a programming language. I don't mind `Set Variable` and I think it both says want it needs to say and no more. So descriptive but short.

**Comment by pekkaklarck** [comment]
For me the biggest issue with `Set Variable` is the inconsistency with `Set Global/Suite/Test Variable`. Having simple syntax that covers all these needs is a win. Still need to make sure we support `Set Variable If` somehow nicely (there are comments about that above). Anyway, `Set Variable` isn't going away in a foreseeable future if ever. It is totally fine to ignore the `VAR` syntax if it is considered complicated.

**Comment by pekkaklarck** [comment]
I've started implementing this and progress has been pretty good. I have the parser part done locally, but I still want to integrate the new functionality to the execution side as well as results before pushing changes.

While implementing this, I have made some decisions related to the open issues discussed above and some new questions have emerged. I'll submit separate questions about each of them.

**Comment by pekkaklarck** [comment]
Decision: I believe we don't need `resolve` (or `resolve_name`) option for controlling should the name be resolved and the resulting value used as the name or not. As others have pointed out above, using `${${name}}` covers the same thing. Nested variables are a somewhat advanced feature, but so is creating a variable name dynamically so this ought to be fine. The option can be added later if needed.

**Comment by pekkaklarck** [comment]
With `${scalar}` variables we need to decide how to handle there more or less values than one:

```
VAR    ${x}    first    second
VAR    ${y}
```

We have the same situation in the Variables section:
```
*** Variables ***
${x}    first    second
${y}
```
The above is handled so that `${x}` gets value `first second` (i.e. values are joined together with a space) and `${y}` becomes an empty string. In addition to that, the separator to be used for joining values can be changed by having a special `SEPARATOR=<sep>` marker as the first value. For example, this creates `${x}` so that it has a multiline value:
```
*** Variables ***
${x}    SEPARATOR=\n
...    First line. This could be long.
...    Second line.
```
This has worked fairly well and also `BuiltIn.Catenate` keyword supports the same syntax. I believe we should use the same approach also with `VAR`, but I think a separate configuration option `separator` works better than making the first value somehow special. In practice that would allow usages like this:
```
VAR    ${message}
...    This is the beginning of the message. This message is rather long.
...    The message continues here.
...    And here. All these parts are joined together with a space.
VAR    ${xml}
...    <root>
...    <child>These parts are joined together with a newline.</child>
...    <child>That makes the resulting XML snippet easier to read.</child>
...    </root>
...    separator=\n
```

The main reason I believe a `separator` configuration option is better than making the first value special is that it is consistent with the `scope` option and also with similar configuration options other control structures and also keywords have.

To allow using the same syntax with `VAR` and in the Variable section, I believe we should support the explicit `separator` option also there. I'll submit a separate issue about that.

**Comment by pekkaklarck** [comment]
One more decision to be made: With the `scope` option we need to decide should we allow variables like `scope=${scope}` or require the value to be specified directly like `scope=suite`. Especially if we don't support variables, we could also decide to require the value be all caps like `scope=SUITE`.

Requiring explicit all caps values would benefit typing (we could use `typing.Literal`). would make it easier for external tools to check what the scope is, and would make it easy for us to validate the scope already in parser. I also don't see use cases for making the scope dynamic by allowing variables.

On the other hand, with other similar config options we support values case-insensitively (e,g, `on_limit` with `WHILE` and `pattern_type` with `EXCEPT) and all options support variables. This option behaving differently would be rather strange and inconsistent.

Although there would be some benefits in making the scope strict, I believe we should allow variables and case-insensitive values for consistency. The main drawback I see is that we cannot directly validate the scope in parser, but I guess we could do it so that validation is done if the value doesn't have variables.

**Comment by MoreFamed** [comment]
> One more decision to be made: With the `scope` option we need to decide should we allow variables like `scope=${scope}` or require the value to be specified directly like `scope=suite`. Especially if we don't support variables, we could also decide to require the value be all caps like `scope=SUITE`.

I would not allow variables for setting the scope. I do not see use case for dynamic setting of the scope. As regards case sensitivity, I have no strong opinion.

**Comment by pekkaklarck** [comment]
Basic functionality is now ready. For example, all these work:
```
VAR    ${name}    value
VAR    ${name}    multiple    values                   # Joined together with a space.
VAR    ${name}    multiple    values    separator=-    # Joined together with '-'.
VAR    @{list}    a    b    c
VAR    &{dict}    key=value    second=item
VAR    ${name}    value    scope=TEST    # Supports GLOBAL, SUITE, TEST, TASK and LOCAL (default).
IF    ${cond}    VAR    ${x}    true value    ELSE    VAR    ${x}    false value
```

**Comment by pekkaklarck** [comment]
These things still need to be done:
- Possibility to create the name dynamically. I'm still not certain is supporting `${${x}}` as the name or adding a separate config option like `resolve_name` better. I was planning to use the former because it works also when using variables. I also thought you can create variables like that in the Variables section, but that wasn't the case. This still needs to be implemented one way or the other.
- Direct support to conditionally create variables. As the examples above demonstrate, `VAR` already now works with inline `IF`, but it requires having both `VAR` and the variable name twice. Something like
   ```
   VAR    ${x}    true value    IF    ${cond}    ELSE    false value
   ```
   would be nice, but I don't consider it a must in RF 7.0.
- Documentation.

**Comment by pekkaklarck** [comment]
I hope users interested in this functionality can test it already now or latest when the first alpha is out. Comments are appreciated here or on the #devel channel in our Slack!

**Comment by pekkaklarck** [comment]
Dynamically creating the name like
```
VAR    ${{x}}    value
```
is not supported. For consistency reasons that support was added also to the Variables section (#4905). The remaining tasks thus are:

- Possible conditional assignment support.
- Documentation.

I'll look the former soonish, but I'm afraid implementing it will be too big a task for RF 7.0. It would effectively add branches the `VAR` syntax making all model objects more complicated.

**Comment by pekkaklarck** [comment]
If we were to support conditional assignment with `VAR`, we needed to specify the syntax. There are at least two alternatives discussed below:

```
VAR    ${x}    y    IF    $cond    ELSE    z
```
This is same structure as with Python's ternary `if/else`. It would thus be familiar for people with Python skills and would also help people moving from Robot syntax to Python.

```
VAR    ${x}    IF    $cond    y    ELSE    z
```
This is close to our inline IF construct. Unlike the above, this would allow `ELSE IF` as well:
```
VAR    ${x}    IF    $c > 0    a    ELSE IF    $c < 0    b    ELSE    c
```
I don't know is `ELSE IF` support really needed, though. It easily gets pretty complicated and would typically be better changed to a normal `IF/ELSE`. Supporting unlimited branches would also be a lot bigger task than adding support for an optional condition and a false value.

Comments about these alternatives and possible new alternatives are appreciated!

**Comment by pekkaklarck** [comment]
The `VAR` syntax has been fine-tuned:

1. It is now possible to have `=` after the variable name the same way as when creating variables based on keyword return values and in the Variables section. This was requested by users on our Slack and implemented in 362581495688492ba7134e059946f75bacdf9c98.
2. Option parsing now accepts each option only once. For example,
   ```
   VAR    &{x}    scope=value    scope=test
   ```
   now creates a variable with value `{'scope': 'value'}` when earlier `scope` was considered to be an option twice which caused an error.

**Comment by pekkaklarck** [comment]
We had a long discussion about the conditional `VAR` syntax on our Slack. We had a vote about syntax alternatives I explained in an [earlier comment](https://github.com/robotframework/robotframework/issues/3761#issuecomment-1770638592) and the 
```
VAR    ${x}    y    IF    $cond    ELSE    z
```
variant was generally preferred.

Things got more complicated when I realized that this kind of conditional assignment could be useful also in the Variables section. In that usage `ELSE IF` support would likely be useful, but the above syntax cannot really be extended to handle it. It could thus be that the
```
VAR    ${x}    IF    $cond    y    ELSE    z
```
variant that would also support `ELSE IF` (see the aforementioned comment for an example) could be a better syntax choice.

**Comment by pekkaklarck** [comment]
In the end it turned out that it is best to post-pone adding the conditional assignment support to some future release. This is why:
* The `VAR` syntax is useful already in its current form.
* It can be used with `IF`, including inline `IF`, to set variables conditionally.
* As I explained in the above comment, it's not clear which syntax would be best.
* Implementing conditional assignment support seems to be considerably more complicated than I initially thought.
* Especially the `ELSE IF` variant would be challenging to implement.
* We are running out of the budget with RF 7.0.
* There are more important features in RF 7.0 scope than conditional `VAR` support.

After this decision the only remaining task related to the `VAR` syntax in RF 7.0 ought to be documentation. Once that's ready, I think we are ready for an alpha release that makes it easier for users to test this new syntax.

**Comment by pekkaklarck** [comment]
Documentation is now done and this issue ought to be ready. Looking forward for feedback once the alpha is out! At this point of the RF 7 development we can still make adjustments to the syntax if needed.

**Comment by MoreFamed** [comment]
> If we were to support conditional assignment with `VAR`, we needed to specify the syntax. There are at least two alternatives discussed below:
> 
> ```
> VAR    ${x}    y    IF    $cond    ELSE    z
> ```
> 
> This is same structure as with Python's ternary `if/else`. It would thus be familiar for people with Python skills and would also help people moving from Robot syntax to Python.
> 
> ```
> VAR    ${x}    IF    $cond    y    ELSE    z
> ```
> 
> This is close to our inline IF construct. Unlike the above, this would allow `ELSE IF` as well:
> 
> ```
> VAR    ${x}    IF    $c > 0    a    ELSE IF    $c < 0    b    ELSE    c
> ```
> 
> I don't know is `ELSE IF` support really needed, though. It easily gets pretty complicated and would typically be better changed to a normal `IF/ELSE`. Supporting unlimited branches would also be a lot bigger task than adding support for an optional condition and a false value.
> 
> Comments about these alternatives and possible new alternatives are appreciated!

I would strongly vote for the second syntax. There are two reasons for this:
- It is more Robotish, since `Set Variable If`and `Run Keyword If` has similar structure (i.e. there always comes the condition before the appropriate value);
- I would use `ELSE IF` support for conditional settings of variables instead of `Set Variable If`.
